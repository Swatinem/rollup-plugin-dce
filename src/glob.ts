import glob from "glob";
import path from "path";

export async function getSourceFiles(pattern: string, options: glob.IOptions) {
  const files: Array<string> = await new Promise((resolve, reject) => {
    glob(pattern, options, (error, matches) => {
      // istanbul ignore if
      if (error) {
        return reject(error);
      }
      resolve(matches);
    });
  });
  return new Set(files.map(f => path.join(options.cwd!, f)));
}
