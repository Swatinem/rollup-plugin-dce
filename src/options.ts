export interface Options {
  /**
   * All file names, and the `sourceGlob` below will be reported and resolved
   * relative to the `context`, which defaults to `process.cwd()`.
   */
  context?: string;
  /**
   * To find files which are not being imported at all, a `sourceGlob` pattern
   * can be provided that will find and report all those unused source files.
   */
  sourceGlob?: string;
}
