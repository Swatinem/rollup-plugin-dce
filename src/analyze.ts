import { OutputAsset, OutputChunk } from "rollup";
import path from "path";
import { getSourceFiles } from "./glob";

interface DeadCodeFile {
  fileName: string;
  deadExports: Array<string>;
}

export interface DeadCode {
  files: Array<DeadCodeFile>;
}

interface Module {
  renderedExports: Set<string>;
  removedExports: Set<string>;
}

interface Options {
  context: string;
  sourceGlob?: string;
  chunks: Array<OutputChunk | OutputAsset>;
}

export async function getDeadCode(options: Options): Promise<DeadCode> {
  const { context, sourceGlob, chunks } = options;
  const sourceFiles = sourceGlob ? await getSourceFiles(sourceGlob, { cwd: context }) : new Set<string>();

  const modules = new Map<string, Module>();

  for (const chunk of chunks) {
    // istanbul ignore if
    if (!chunk.code) {
      continue;
    }
    for (const [id, module] of Object.entries(chunk.modules)) {
      const existingModule = modules.get(id);
      if (existingModule) {
        for (const removedExport of module.removedExports) {
          if (!existingModule.renderedExports.has(removedExport)) {
            existingModule.removedExports.add(removedExport);
          }
        }
        for (const renderedExport of module.renderedExports) {
          existingModule.renderedExports.add(renderedExport);
          existingModule.removedExports.delete(renderedExport);
        }
      } else {
        modules.set(id, {
          renderedExports: new Set(module.renderedExports),
          removedExports: new Set(module.removedExports),
        });
      }
    }
  }

  const files: Array<DeadCodeFile> = [];

  for (const [id, module] of modules) {
    sourceFiles.delete(id);
    if (module.removedExports.size) {
      files.push({
        fileName: path.relative(context, id),
        deadExports: [...module.removedExports],
      });
    }
  }

  for (const id of sourceFiles) {
    files.push({
      fileName: path.relative(context, id),
      deadExports: [],
    });
  }

  return { files };
}
