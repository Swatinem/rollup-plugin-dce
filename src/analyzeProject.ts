import { Options as PluginOptions } from "./options";
import { rollup, InputOptions } from "rollup";
import { getDeadCode, DeadCode } from "./analyze";

interface Options extends InputOptions, PluginOptions {}

export async function analyzeTSProject(options: Options): Promise<DeadCode> {
  const { context = process.cwd(), sourceGlob, ...rollupOptions } = options;
  const { ts, dts } = await import("rollup-plugin-dts");

  const inputOptions: InputOptions = {
    external(id) {
      // for now all bare imports are considered externalâ€¦
      return !id.startsWith(".") && !id.startsWith("/");
    },
    ...rollupOptions,
    onwarn() {},
  };

  // run once with `ts`
  let bundle = await rollup({
    ...inputOptions,
    plugins: [ts()],
  });
  let result = await bundle.generate({ format: "es" });
  const chunks = result.output;

  // and again with `dts`
  bundle = await rollup({
    ...inputOptions,
    plugins: [dts()],
  });
  result = await bundle.generate({ format: "es" });
  chunks.push(...result.output);

  const deadCode = getDeadCode({
    context,
    sourceGlob,
    chunks,
  });

  return deadCode;
}
