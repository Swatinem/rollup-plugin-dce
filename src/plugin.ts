import { PluginImpl } from "rollup";
import { getDeadCode } from "./analyze";
import { Options } from "./options";

export const plugin: PluginImpl<Options> = (options = {}) => {
  return {
    name: "dce",
    async generateBundle(_options, bundle) {
      const { context = process.cwd(), sourceGlob } = options;

      const { files } = await getDeadCode({ context, sourceGlob, chunks: Object.values(bundle) });
      if (files.length) {
        let formattedError = `Found Dead Code:\n`;
        formattedError += files.map(
          ({ fileName, deadExports }) => `* ${fileName}${deadExports.length ? `\n  ${deadExports.join(", ")}` : ""}`,
        );
        this.warn(formattedError);
      }
    },
  };
};
