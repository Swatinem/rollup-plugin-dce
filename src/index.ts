import { plugin } from "./plugin";
export { analyzeTSProject } from "./analyzeProject";
export { plugin };

export default plugin;
