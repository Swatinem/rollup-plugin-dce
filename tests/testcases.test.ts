import { InputOption } from "rollup";
import { analyzeTSProject } from "../src";
import path from "path";
import fsExtra from "fs-extra";

const TESTCASES = path.join(__dirname, "testcases");

interface Meta {
  skip: boolean;
  cwd: string;
  input: InputOption;
  sourceGlob: string;
}

describe("testcases", () => {
  const cwd = process.cwd();
  afterAll(() => process.chdir(cwd));

  const dirs = fsExtra.readdirSync(TESTCASES);
  for (const name of dirs) {
    const dir = path.join(TESTCASES, name);

    if (fsExtra.statSync(dir).isDirectory()) {
      const config: Meta = {
        cwd: dir,
        skip: false,
        input: "index.js",
        sourceGlob: "!(config).{ts,js}",
      };

      try {
        const configFile = require(path.join(dir, "config"));
        Object.assign(config, configFile.default || configFile);
      } catch {}

      const testfn = config.skip ? it.skip : it;

      testfn(`works for testcase "${name}"`, async () => {
        process.chdir(config.cwd);

        await expect(
          analyzeTSProject({
            input: config.input,
            sourceGlob: config.sourceGlob,
          }),
        ).resolves.toMatchSnapshot();
      });
    }
  }
});
