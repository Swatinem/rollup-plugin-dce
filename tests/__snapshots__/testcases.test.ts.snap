// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testcases works for testcase "dead-file" 1`] = `
Object {
  "files": Array [
    Object {
      "deadExports": Array [],
      "fileName": "a.js",
    },
  ],
}
`;

exports[`testcases works for testcase "dts" 1`] = `
Object {
  "files": Array [
    Object {
      "deadExports": Array [
        "C",
      ],
      "fileName": "common.ts",
    },
  ],
}
`;

exports[`testcases works for testcase "export-statements" 1`] = `
Object {
  "files": Array [
    Object {
      "deadExports": Array [
        "b",
      ],
      "fileName": "a.js",
    },
  ],
}
`;

exports[`testcases works for testcase "external" 1`] = `
Object {
  "files": Array [],
}
`;

exports[`testcases works for testcase "multi-entry" 1`] = `
Object {
  "files": Array [
    Object {
      "deadExports": Array [
        "c",
      ],
      "fileName": "common.js",
    },
  ],
}
`;

exports[`testcases works for testcase "namespace-exports" 1`] = `
Object {
  "files": Array [
    Object {
      "deadExports": Array [
        "b",
      ],
      "fileName": "namespace.js",
    },
    Object {
      "deadExports": Array [
        "ns",
      ],
      "fileName": "reexport.js",
    },
  ],
}
`;

exports[`testcases works for testcase "non-exported" 1`] = `
Object {
  "files": Array [
    Object {
      "deadExports": Array [
        "b",
      ],
      "fileName": "a.js",
    },
  ],
}
`;
